src/fundus/scraping/pipeline.py

```diff
@@ -62,6 +63,104 @@ def __init__(self, *scrapers: Scraper):
+        url_filter: Optional[URLFilter] = None,
+        only_unique: bool = True,
+    ) -> AsyncIterator[Article]:
+        """Async variant of the crawl() method.
+
+        See docstring for crawl(). for detailed information about the parameters.
+
+        Args:
+            max_articles (Optional[int]): Number of articles to crawl. Defaults to None.
+            error_handling (Literal["suppress", "catch", "raise"]): Set error handling. Defaults to "suppress".
+            only_complete (Union[bool, ExtractionFilter]): Set extraction filters. Defaults to True
+            delay (Optional[Union[float, Delay]]): Set delay time between article batches. Defaults to None.
+            url_filter (Optional[URLFilter]): Set URLFilter. Defaults to None.
+            only_unique (bool): If true return only unique responses. Defaults to True.
+
+        Returns:
+            AsyncIterator[Article]: An iterator yielding objects of type Article.
+        """
+
+        def build_extraction_filter() -> Optional[ExtractionFilter]:
+            if isinstance(only_complete, bool):
+                return (
+                    None
+                    if only_complete is False
+                    else lambda extracted: not all(
+                        bool(v) if not isinstance(v, Exception) else False for _, v in extracted.items()
+                    )
+                )
+            else:
+                return only_complete
+
+        def build_delay() -> Optional[Delay]:
+            if isinstance(delay, float):
+
+                def constant_delay() -> float:
+                    return delay  # type: ignore[return-value]
+
+                return constant_delay
+            else:
+                return delay
+
+        def build_unique_url_filter() -> URLFilter:
+            return lambda url: url in response_cache
+
+        # build filters and delay. this is for readability and typeguard reasons
+        extraction_filter = build_extraction_filter()
+        unique_url_filter = build_unique_url_filter() if only_unique else None
+        final_delay = build_delay()
+
+        response_cache: Set[str] = set()
+
+        for scraper in self.scrapers:
+            for source in scraper.sources:
+                if url_filter:
+                    source.add_url_filter(url_filter=url_filter)
+                if unique_url_filter:
+                    source.add_url_filter(url_filter=unique_url_filter)
+
+        async_article_iterators: List[AsyncIterator[Optional[Article]]] = [
+            scraper.scrape(
+                error_handling=error_handling,
+                extraction_filter=extraction_filter,
+            )
+            for scraper in self.scrapers
+        ]
+
+        # we use this custom variant of interleave_longest in order to be able
+        # to delay the program flow between batches
+        async def _async_article_interleave_longest() -> AsyncIterator[Article]:
+            # type: ignore[attr-defined]
```

Comment:
This is properly a leftover type-ignore.

Reply:
Yes, i removed it.